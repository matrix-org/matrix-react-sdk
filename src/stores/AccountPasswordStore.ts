/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

const PASSWORD_TIMEOUT = 5 * 60 * 1000; // five minutes

/**
 * Store for the account password.
 * This password can be used for a short time after login
 * to avoid requestin the password all the time for instance during e2ee setup.
 */
export class AccountPasswordStore {
    private password?: string;
    private passwordTimeoutId?: ReturnType<typeof setTimeout>;

    public setPassword(password: string): void {
        this.password = password;
        clearTimeout(this.passwordTimeoutId);
        this.passwordTimeoutId = setTimeout(this.clearPassword, PASSWORD_TIMEOUT);
    }

    public getPassword(): string | undefined {
        return this.password;
    }

    public clearPassword = (): void => {
        clearTimeout(this.passwordTimeoutId);
        this.passwordTimeoutId = undefined;
        this.password = undefined;
    };
}
