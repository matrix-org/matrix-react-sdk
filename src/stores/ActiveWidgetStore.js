/*
Copyright 2018 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import EventEmitter from 'events';

import {MatrixClientPeg} from '../MatrixClientPeg';

/**
 * Stores information about the widgets active in the app right now:
 *  * What widget is set to remain always-on-screen, if any
 *    Only one widget may be 'always on screen' at any one time.
 *  * Negotiated capabilities for active apps
 */
class ActiveWidgetStore extends EventEmitter {
    constructor() {
        super();
        this._persistentWidgetId = null;

        // A list of negotiated capabilities for each widget, by ID
        // {
        //     widgetId: [caps...],
        // }
        this._capsByWidgetId = {};

        // A WidgetMessaging instance for each widget ID
        this._widgetMessagingByWidgetId = {};

        // What room ID each widget is associated with (if it's a room widget)
        this._roomIdByWidgetId = {};

        this.onRoomStateEvents = this.onRoomStateEvents.bind(this);

        this.dispatcherRef = null;
    }

    start() {
        MatrixClientPeg.get().on('RoomState.events', this.onRoomStateEvents);
    }

    stop() {
        if (MatrixClientPeg.get()) {
            MatrixClientPeg.get().removeListener('RoomState.events', this.onRoomStateEvents);
        }
        this._capsByWidgetId = {};
        this._widgetMessagingByWidgetId = {};
        this._roomIdByWidgetId = {};
    }

    onRoomStateEvents(ev, state) {
        // XXX: This listens for state events in order to remove the active widget.
        // Everything else relies on views listening for events and calling setters
        // on this class which is terrible. This store should just listen for events
        // and keep itself up to date.
        // TODO: Enable support for m.widget event type (https://github.com/vector-im/riot-web/issues/13111)
        if (ev.getType() !== 'im.vector.modular.widgets') return;

        if (ev.getStateKey() === this._persistentWidgetId) {
            this.destroyPersistentWidget(this._persistentWidgetId);
        }
    }

    destroyPersistentWidget(id) {
        if (id !== this._persistentWidgetId) return;
        const toDeleteId = this._persistentWidgetId;

        this.setWidgetPersistence(toDeleteId, false);
        this.delWidgetMessaging(toDeleteId);
        this.delWidgetCapabilities(toDeleteId);
        this.delRoomId(toDeleteId);
    }

    setWidgetPersistence(widgetId, val) {
        if (this._persistentWidgetId === widgetId && !val) {
            this._persistentWidgetId = null;
        } else if (this._persistentWidgetId !== widgetId && val) {
            this._persistentWidgetId = widgetId;
        }
        this.emit('update');
    }

    getWidgetPersistence(widgetId) {
        return this._persistentWidgetId === widgetId;
    }

    getPersistentWidgetId() {
        return this._persistentWidgetId;
    }

    setWidgetCapabilities(widgetId, caps) {
        this._capsByWidgetId[widgetId] = caps;
        this.emit('update');
    }

    widgetHasCapability(widgetId, cap) {
        return this._capsByWidgetId[widgetId] && this._capsByWidgetId[widgetId].includes(cap);
    }

    delWidgetCapabilities(widgetId) {
        delete this._capsByWidgetId[widgetId];
        this.emit('update');
    }

    setWidgetMessaging(widgetId, wm) {
        // Stop any existing widget messaging first
        this.delWidgetMessaging(widgetId);
        this._widgetMessagingByWidgetId[widgetId] = wm;
        this.emit('update');
    }

    getWidgetMessaging(widgetId) {
        return this._widgetMessagingByWidgetId[widgetId];
    }

    delWidgetMessaging(widgetId) {
        if (this._widgetMessagingByWidgetId[widgetId]) {
            try {
                this._widgetMessagingByWidgetId[widgetId].stop();
            } catch (e) {
                console.error('Failed to stop listening for widgetMessaging events', e.message);
            }
            delete this._widgetMessagingByWidgetId[widgetId];
            this.emit('update');
        }
    }

    getRoomId(widgetId) {
        return this._roomIdByWidgetId[widgetId];
    }

    setRoomId(widgetId, roomId) {
        this._roomIdByWidgetId[widgetId] = roomId;
        this.emit('update');
    }

    delRoomId(widgetId) {
        delete this._roomIdByWidgetId[widgetId];
        this.emit('update');
    }
}

if (global.singletonActiveWidgetStore === undefined) {
    global.singletonActiveWidgetStore = new ActiveWidgetStore();
}
export default global.singletonActiveWidgetStore;
