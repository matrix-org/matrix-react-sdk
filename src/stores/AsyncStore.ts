/*
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { EventEmitter } from 'events';
import AwaitLock from 'await-lock';
import { Dispatcher } from "flux";
import { ActionPayload } from "../dispatcher/payloads";

/**
 * The event/channel to listen for in an AsyncStore.
 */
export const UPDATE_EVENT = "update";

/**
 * Represents a minimal store which works similar to Flux stores. Instead
 * of everything needing to happen in a dispatch cycle, everything can
 * happen async to that cycle.
 *
 * The store operates by using Object.assign() to mutate state - it sends the
 * state objects (current and new) through the function onto a new empty
 * object. Because of this, it is recommended to break out your state to be as
 * safe as possible. The state mutations are also locked, preventing concurrent
 * writes.
 *
 * All updates to the store happen on the UPDATE_EVENT event channel with the
 * one argument being the instance of the store.
 *
 * To update the state, use updateState() and preferably await the result to
 * help prevent lock conflicts.
 */
export abstract class AsyncStore<T extends Object> extends EventEmitter {
    private storeState: Readonly<T>;
    private lock = new AwaitLock();
    private readonly dispatcherRef: string;

    /**
     * Creates a new AsyncStore using the given dispatcher.
     * @param {Dispatcher<ActionPayload>} dispatcher The dispatcher to rely upon.
     * @param {T} initialState The initial state for the store.
     */
    protected constructor(private dispatcher: Dispatcher<ActionPayload>, initialState: T = <T>{}) {
        super();

        this.dispatcherRef = dispatcher.register(this.onDispatch.bind(this));
        this.storeState = initialState;
    }

    /**
     * The current state of the store. Cannot be mutated.
     */
    protected get state(): T {
        return this.storeState;
    }

    /**
     * Stops the store's listening functions, such as the listener to the dispatcher.
     */
    protected stop() {
        if (this.dispatcherRef) this.dispatcher.unregister(this.dispatcherRef);
    }

    /**
     * Updates the state of the store.
     * @param {T|*} newState The state to update in the store using Object.assign()
     */
    protected async updateState(newState: T | Object) {
        await this.lock.acquireAsync();
        try {
            this.storeState = Object.freeze(Object.assign(<T>{}, this.storeState, newState));
            this.emit(UPDATE_EVENT, this);
        } finally {
            await this.lock.release();
        }
    }

    /**
     * Resets the store's to the provided state or an empty object.
     * @param {T|*} newState The new state of the store.
     * @param {boolean} quiet If true, the function will not raise an UPDATE_EVENT.
     */
    protected async reset(newState: T | Object = null, quiet = false) {
        await this.lock.acquireAsync();
        try {
            this.storeState = Object.freeze(<T>(newState || {}));
            if (!quiet) this.emit(UPDATE_EVENT, this);
        } finally {
            await this.lock.release();
        }
    }

    /**
     * Called when the dispatcher broadcasts a dispatch event.
     * @param {ActionPayload} payload The event being dispatched.
     */
    protected abstract onDispatch(payload: ActionPayload);
}
