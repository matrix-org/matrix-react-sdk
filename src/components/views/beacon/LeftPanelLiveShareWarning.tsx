/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import classNames from 'classnames';
import React from 'react';

import { useEventEmitterState } from '../../../hooks/useEventEmitter';
import { _t } from '../../../languageHandler';
import { OwnBeaconStore, OwnBeaconStoreEvent } from '../../../stores/OwnBeaconStore';
import { Icon as LiveLocationIcon } from '../../../../res/img/location/live-location.svg';
import { ViewRoomPayload } from '../../../dispatcher/payloads/ViewRoomPayload';
import { Action } from '../../../dispatcher/actions';
import dispatcher from '../../../dispatcher/dispatcher';
import AccessibleButton from '../elements/AccessibleButton';

interface Props {
    isMinimized?: boolean;
}

/**
 * Choose the most relevant beacon
 * and get its roomId
 */
const chooseBestBeaconRoomId = (liveBeaconIds, errorBeaconIds): string | undefined => {
    // both lists are ordered by creation timestamp in store
    // so select latest beacon
    const beaconId = errorBeaconIds?.[0] ?? liveBeaconIds?.[0];
    if (!beaconId) {
        return undefined;
    }
    const beacon = OwnBeaconStore.instance.getBeaconById(beaconId);

    return beacon?.roomId;
};

const LeftPanelLiveShareWarning: React.FC<Props> = ({ isMinimized }) => {
    const isMonitoringLiveLocation = useEventEmitterState(
        OwnBeaconStore.instance,
        OwnBeaconStoreEvent.MonitoringLivePosition,
        () => OwnBeaconStore.instance.isMonitoringLiveLocation,
    );

    const beaconIdsWithWireError = useEventEmitterState(
        OwnBeaconStore.instance,
        OwnBeaconStoreEvent.WireError,
        () => OwnBeaconStore.instance.getLiveBeaconIdsWithWireError(),
    );

    const liveBeaconIds = useEventEmitterState(
        OwnBeaconStore.instance,
        OwnBeaconStoreEvent.LivenessChange,
        () => OwnBeaconStore.instance.getLiveBeaconIds(),
    );

    const hasWireErrors = !!beaconIdsWithWireError.length;

    if (!isMonitoringLiveLocation) {
        return null;
    }

    const relevantBeaconRoomId = chooseBestBeaconRoomId(liveBeaconIds, beaconIdsWithWireError);

    const onWarningClick = relevantBeaconRoomId ? () => {
        dispatcher.dispatch<ViewRoomPayload>({
            action: Action.ViewRoom,
            room_id: relevantBeaconRoomId,
            metricsTrigger: undefined,
        });
    } : undefined;

    const label = hasWireErrors ?
        _t('An error occured whilst sharing your live location') :
        _t('You are sharing your live location');

    return <AccessibleButton
        className={classNames('mx_LeftPanelLiveShareWarning', {
            'mx_LeftPanelLiveShareWarning__minimized': isMinimized,
            'mx_LeftPanelLiveShareWarning__error': hasWireErrors,
        })}
        title={isMinimized ? label : undefined}
        onClick={onWarningClick}
    >
        { isMinimized ? <LiveLocationIcon height={10} /> : label }
    </AccessibleButton>;
};

export default LeftPanelLiveShareWarning;
