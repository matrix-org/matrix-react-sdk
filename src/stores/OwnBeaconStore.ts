/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { debounce } from "lodash";
import {
    Beacon,
    BeaconEvent,
    MatrixEvent,
    Room,
    RoomMember,
    RoomState,
    RoomStateEvent,
} from "matrix-js-sdk/src/matrix";
import {
    BeaconInfoState, makeBeaconContent, makeBeaconInfoContent,
} from "matrix-js-sdk/src/content-helpers";
import { M_BEACON } from "matrix-js-sdk/src/@types/beacon";
import { logger } from "matrix-js-sdk/src/logger";

import defaultDispatcher from "../dispatcher/dispatcher";
import { ActionPayload } from "../dispatcher/payloads";
import { AsyncStoreWithClient } from "./AsyncStoreWithClient";
import { arrayDiff } from "../utils/arrays";
import {
    ClearWatchCallback,
    GeolocationError,
    mapGeolocationPositionToTimedGeo,
    sortBeaconsByLatestCreation,
    TimedGeoUri,
    watchPosition,
} from "../utils/beacon";
import { getCurrentPosition } from "../utils/beacon/geolocation";

const isOwnBeacon = (beacon: Beacon, userId: string): boolean => beacon.beaconInfoOwner === userId;

export enum OwnBeaconStoreEvent {
    LivenessChange = 'OwnBeaconStore.LivenessChange',
    MonitoringLivePosition = 'OwnBeaconStore.MonitoringLivePosition',
    WireError = 'WireError',
}

const MOVING_UPDATE_INTERVAL = 2000;
const STATIC_UPDATE_INTERVAL = 30000;

const BAIL_AFTER_CONSECUTIVE_ERROR_COUNT = 2;

type OwnBeaconStoreState = {
    beacons: Map<string, Beacon>;
    beaconWireErrors: Map<string, Beacon>;
    beaconsByRoomId: Map<Room['roomId'], Set<string>>;
    liveBeaconIds: string[];
};
export class OwnBeaconStore extends AsyncStoreWithClient<OwnBeaconStoreState> {
    private static internalInstance = new OwnBeaconStore();
    // users beacons, keyed by event type
    public readonly beacons = new Map<string, Beacon>();
    public readonly beaconsByRoomId = new Map<Room['roomId'], Set<string>>();
    /**
     * Track over the wire errors for published positions
     * Counts consecutive wire errors per beacon
     * Reset on successful publish of location
     */
    public readonly beaconWireErrorCounts = new Map<string, number>();
    /**
     * ids of live beacons
     * ordered by creation time descending
     */
    private liveBeaconIds = [];
    private locationInterval: number;
    private geolocationError: GeolocationError | undefined;
    private clearPositionWatch: ClearWatchCallback | undefined;
    /**
     * Track when the last position was published
     * So we can manually get position on slow interval
     * when the target is stationary
     */
    private lastPublishedPositionTimestamp: number | undefined;

    public constructor() {
        super(defaultDispatcher);
    }

    public static get instance(): OwnBeaconStore {
        return OwnBeaconStore.internalInstance;
    }

    /**
     * True when we have live beacons
     * and geolocation.watchPosition is active
     */
    public get isMonitoringLiveLocation(): boolean {
        return !!this.clearPositionWatch;
    }

    protected async onNotReady() {
        this.matrixClient.removeListener(BeaconEvent.LivenessChange, this.onBeaconLiveness);
        this.matrixClient.removeListener(BeaconEvent.New, this.onNewBeacon);
        this.matrixClient.removeListener(RoomStateEvent.Members, this.onRoomStateMembers);

        this.beacons.forEach(beacon => beacon.destroy());

        this.stopPollingLocation();
        this.beacons.clear();
        this.beaconsByRoomId.clear();
        this.liveBeaconIds = [];
        this.beaconWireErrorCounts.clear();
    }

    protected async onReady(): Promise<void> {
        this.matrixClient.on(BeaconEvent.LivenessChange, this.onBeaconLiveness);
        this.matrixClient.on(BeaconEvent.New, this.onNewBeacon);
        this.matrixClient.on(RoomStateEvent.Members, this.onRoomStateMembers);

        this.initialiseBeaconState();
    }

    protected async onAction(payload: ActionPayload): Promise<void> {
        // we don't actually do anything here
    }

    public hasLiveBeacons = (roomId?: string): boolean => {
        return !!this.getLiveBeaconIds(roomId).length;
    };

    /**
     * Some live beacon has a wire error
     * Optionally filter by room
     */
    public hasWireErrors = (roomId?: string): boolean => {
        return this.getLiveBeaconIds(roomId).some(this.beaconHasWireError);
    };

    /**
     * If a beacon has failed to publish position
     * past the allowed consecutive failure count (BAIL_AFTER_CONSECUTIVE_ERROR_COUNT)
     * Then consider it to have an error
     */
    public beaconHasWireError = (beaconId: string): boolean => {
        return this.beaconWireErrorCounts.get(beaconId) >= BAIL_AFTER_CONSECUTIVE_ERROR_COUNT;
    };

    public resetWireError = (beaconId: string): void => {
        this.incrementBeaconWireErrorCount(beaconId, false);

        // always publish to all live beacons together
        // instead of just one that was changed
        // to keep lastPublishedTimestamp simple
        // and extra published locations don't hurt
        this.publishCurrentLocationToBeacons();
    };

    public getLiveBeaconIds = (roomId?: string): string[] => {
        if (!roomId) {
            return this.liveBeaconIds;
        }
        return this.liveBeaconIds.filter(beaconId => this.beaconsByRoomId.get(roomId)?.has(beaconId));
    };

    public getLiveBeaconIdsWithWireError = (roomId?: string): string[] => {
        return this.getLiveBeaconIds(roomId).filter(this.beaconHasWireError);
    };

    public getBeaconById = (beaconId: string): Beacon | undefined => {
        return this.beacons.get(beaconId);
    };

    public stopBeacon = async (beaconInfoType: string): Promise<void> => {
        const beacon = this.beacons.get(beaconInfoType);
        // if no beacon, or beacon is already explicitly set isLive: false
        // do nothing
        if (!beacon?.beaconInfo?.live) {
            return;
        }

        return await this.updateBeaconEvent(beacon, { live: false });
    };

    /**
     * Listeners
     */

    private onNewBeacon = (_event: MatrixEvent, beacon: Beacon): void => {
        if (!isOwnBeacon(beacon, this.matrixClient.getUserId())) {
            return;
        }
        this.addBeacon(beacon);
        this.checkLiveness();
    };

    private onBeaconLiveness = (isLive: boolean, beacon: Beacon): void => {
        // check if we care about this beacon
        if (!this.beacons.has(beacon.identifier)) {
            return;
        }

        // beacon expired, update beacon to un-alive state
        if (!isLive) {
            this.stopBeacon(beacon.identifier);
        }

        this.checkLiveness();

        this.emit(OwnBeaconStoreEvent.LivenessChange, this.getLiveBeaconIds());
    };

    /**
     * Check for changes in membership in rooms with beacons
     * and stop monitoring beacons in rooms user is no longer member of
     */
    private onRoomStateMembers = (_event: MatrixEvent, roomState: RoomState, member: RoomMember): void => {
        // no beacons for this room, ignore
        if (
            !this.beaconsByRoomId.has(roomState.roomId) ||
            member.userId !== this.matrixClient.getUserId()
        ) {
            return;
        }

        // TODO check powerlevels here
        // in PSF-797

        // stop watching beacons in rooms where user is no longer a member
        if (member.membership === 'leave' || member.membership === 'ban') {
            this.beaconsByRoomId.get(roomState.roomId).forEach(this.removeBeacon);
            this.beaconsByRoomId.delete(roomState.roomId);
        }
    };

    /**
     * State management
     */

    /**
     * Live beacon ids that do not have wire errors
     */
    private get healthyLiveBeaconIds() {
        return this.liveBeaconIds.filter(beaconId => !this.beaconHasWireError(beaconId));
    }

    private initialiseBeaconState = () => {
        const userId = this.matrixClient.getUserId();
        const visibleRooms = this.matrixClient.getVisibleRooms();

        visibleRooms
            .forEach(room => {
                const roomState = room.currentState;
                const beacons = roomState.beacons;
                const ownBeaconsArray = [...beacons.values()].filter(beacon => isOwnBeacon(beacon, userId));
                ownBeaconsArray.forEach(beacon => this.addBeacon(beacon));
            });

        this.checkLiveness();
    };

    private addBeacon = (beacon: Beacon): void => {
        this.beacons.set(beacon.identifier, beacon);

        if (!this.beaconsByRoomId.has(beacon.roomId)) {
            this.beaconsByRoomId.set(beacon.roomId, new Set<string>());
        }

        this.beaconsByRoomId.get(beacon.roomId).add(beacon.identifier);

        beacon.monitorLiveness();
    };

    /**
     * Remove listeners for a given beacon
     * remove from state
     * and update liveness if changed
     */
    private removeBeacon = (beaconId: string): void => {
        if (!this.beacons.has(beaconId)) {
            return;
        }
        this.beacons.get(beaconId).destroy();
        this.beacons.delete(beaconId);

        this.checkLiveness();
    };

    private checkLiveness = (): void => {
        const prevLiveBeaconIds = this.getLiveBeaconIds();
        this.liveBeaconIds = [...this.beacons.values()]
            .filter(beacon => beacon.isLive)
            .sort(sortBeaconsByLatestCreation)
            .map(beacon => beacon.identifier);

        const diff = arrayDiff(prevLiveBeaconIds, this.liveBeaconIds);

        if (diff.added.length || diff.removed.length) {
            this.emit(OwnBeaconStoreEvent.LivenessChange, this.liveBeaconIds);
        }

        // publish current location immediately
        // when there are new live beacons
        // and we already have a live monitor
        // so first position is published quickly
        // even when target is stationary
        //
        // when there is no existing live monitor
        // it will be created below by togglePollingLocation
        // and publish first position quickly
        if (diff.added.length && this.isMonitoringLiveLocation) {
            this.publishCurrentLocationToBeacons();
        }

        // if overall liveness changed
        if (!!prevLiveBeaconIds?.length !== !!this.liveBeaconIds.length) {
            this.togglePollingLocation();
        }
    };

    /**
     * Geolocation
     */

    private togglePollingLocation = () => {
        if (!!this.liveBeaconIds.length) {
            this.startPollingLocation();
        } else {
            this.stopPollingLocation();
        }
    };

    private startPollingLocation = async () => {
        // clear any existing interval
        this.stopPollingLocation();

        try {
            this.clearPositionWatch = await watchPosition(
                this.onWatchedPosition,
                this.onGeolocationError,
            );
        } catch (error) {
            this.onGeolocationError(error?.message);
            // don't set locationInterval if geolocation failed to setup
            return;
        }

        this.locationInterval = setInterval(() => {
            if (!this.lastPublishedPositionTimestamp) {
                return;
            }
            // if position was last updated STATIC_UPDATE_INTERVAL ms ago or more
            // get our position and publish it
            if (this.lastPublishedPositionTimestamp <= Date.now() - STATIC_UPDATE_INTERVAL) {
                this.publishCurrentLocationToBeacons();
            }
        }, STATIC_UPDATE_INTERVAL);

        this.emit(OwnBeaconStoreEvent.MonitoringLivePosition);
    };

    private stopPollingLocation = () => {
        clearInterval(this.locationInterval);
        this.locationInterval = undefined;
        this.lastPublishedPositionTimestamp = undefined;
        this.geolocationError = undefined;

        if (this.clearPositionWatch) {
            this.clearPositionWatch();
            this.clearPositionWatch = undefined;
        }

        this.emit(OwnBeaconStoreEvent.MonitoringLivePosition);
    };

    private onWatchedPosition = (position: GeolocationPosition) => {
        const timedGeoPosition = mapGeolocationPositionToTimedGeo(position);

        // if this is our first position, publish immediateley
        if (!this.lastPublishedPositionTimestamp) {
            this.publishLocationToBeacons(timedGeoPosition);
        } else {
            this.debouncedPublishLocationToBeacons(timedGeoPosition);
        }
    };

    private onGeolocationError = async (error: GeolocationError): Promise<void> => {
        this.geolocationError = error;
        logger.error('Geolocation failed', this.geolocationError);

        // other errors are considered non-fatal
        // and self recovering
        if (![
            GeolocationError.Unavailable,
            GeolocationError.PermissionDenied,
        ].includes(error)) {
            return;
        }

        this.stopPollingLocation();
        // kill live beacons when location permissions are revoked
        // TODO may need adjustment when PSF-797 is done
        await Promise.all(this.liveBeaconIds.map(this.stopBeacon));
    };

    /**
     * Gets the current location
     * (as opposed to using watched location)
     * and publishes it to all live beacons
     */
    private publishCurrentLocationToBeacons = async () => {
        try {
            const position = await getCurrentPosition();
            this.publishLocationToBeacons(mapGeolocationPositionToTimedGeo(position));
        } catch (error) {
            this.onGeolocationError(error?.message);
        }
    };

    /**
     * MatrixClient api
     */

    private updateBeaconEvent = async (beacon: Beacon, update: Partial<BeaconInfoState>): Promise<void> => {
        const { description, timeout, timestamp, live, assetType } = {
            ...beacon.beaconInfo,
            ...update,
        };

        const updateContent = makeBeaconInfoContent(timeout,
            live,
            description,
            assetType,
            timestamp);

        await this.matrixClient.unstable_setLiveBeacon(beacon.roomId, beacon.beaconInfoEventType, updateContent);
    };

    /**
     * Sends m.location events to all live beacons
     * Sets last published beacon
     */
    private publishLocationToBeacons = async (position: TimedGeoUri) => {
        this.lastPublishedPositionTimestamp = Date.now();
        await Promise.all(this.healthyLiveBeaconIds.map(beaconId =>
            this.sendLocationToBeacon(this.beacons.get(beaconId), position)),
        );
    };

    private debouncedPublishLocationToBeacons = debounce(this.publishLocationToBeacons, MOVING_UPDATE_INTERVAL);

    /**
     * Sends m.location event to referencing given beacon
     */
    private sendLocationToBeacon = async (beacon: Beacon, { geoUri, timestamp }: TimedGeoUri) => {
        const content = makeBeaconContent(geoUri, timestamp, beacon.beaconInfoId);
        try {
            await this.matrixClient.sendEvent(beacon.roomId, M_BEACON.name, content);
            this.incrementBeaconWireErrorCount(beacon.identifier, false);
        } catch (error) {
            logger.error(error);
            this.incrementBeaconWireErrorCount(beacon.identifier, true);
        }
    };

    /**
     * Manage beacon wire error count
     * - clear count for beacon when not error
     * - increment count for beacon when is error
     * - emit if beacon error count crossed threshold
     */
    private incrementBeaconWireErrorCount = (beaconId: string, isError: boolean): void => {
        const hadError = this.beaconHasWireError(beaconId);

        if (isError) {
            // increment error count
            this.beaconWireErrorCounts.set(
                beaconId,
                (this.beaconWireErrorCounts.get(beaconId) ?? 0) + 1,
            );
        } else {
            // clear any error count
            this.beaconWireErrorCounts.delete(beaconId);
        }

        if (this.beaconHasWireError(beaconId) !== hadError) {
            this.emit(OwnBeaconStoreEvent.WireError, beaconId);
        }
    };
}
