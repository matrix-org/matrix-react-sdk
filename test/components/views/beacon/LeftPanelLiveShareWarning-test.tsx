/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { mocked } from 'jest-mock';
import { mount } from 'enzyme';

import '../../../skinned-sdk';
import LeftPanelLiveShareWarning from '../../../../src/components/views/beacon/LeftPanelLiveShareWarning';
import { OwnBeaconStore, OwnBeaconStoreEvent } from '../../../../src/stores/OwnBeaconStore';
import { flushPromises } from '../../../test-utils';

jest.mock('../../../../src/stores/OwnBeaconStore', () => {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const EventEmitter = require("events");
    class MockOwnBeaconStore extends EventEmitter {
        public hasLiveBeacons = jest.fn().mockReturnValue(false);
    }
    return {
        // @ts-ignore
        ...jest.requireActual('../../../../src/stores/OwnBeaconStore'),
        OwnBeaconStore: {
            instance: new MockOwnBeaconStore() as unknown as OwnBeaconStore,
        },
    };
},
);

describe('<LeftPanelLiveShareWarning />', () => {
    const defaultProps = {};
    const getComponent = (props = {}) =>
        mount(<LeftPanelLiveShareWarning {...defaultProps} {...props} />);

    it('renders nothing when user has no live beacons', () => {
        const component = getComponent();
        expect(component.html()).toBe(null);
    });

    describe('when user has live beacons', () => {
        beforeEach(() => {
            mocked(OwnBeaconStore.instance).hasLiveBeacons.mockReturnValue(true);
        });
        it('renders correctly when not minimized', () => {
            const component = getComponent();
            expect(component).toMatchSnapshot();
        });

        it('renders correctly when minimized', () => {
            const component = getComponent({ isMinimized: true });
            expect(component).toMatchSnapshot();
        });

        it('removes itself when user stops having live beacons', async () => {
            const component = getComponent({ isMinimized: true });
            // started out rendered
            expect(component.html()).toBeTruthy();

            mocked(OwnBeaconStore.instance).hasLiveBeacons.mockReturnValue(false);
            OwnBeaconStore.instance.emit(OwnBeaconStoreEvent.LivenessChange, false);

            await flushPromises();
            component.setProps({});

            expect(component.html()).toBe(null);
        });
    });
});
