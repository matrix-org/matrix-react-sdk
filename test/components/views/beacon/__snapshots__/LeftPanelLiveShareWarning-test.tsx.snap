// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LeftPanelLiveShareWarning /> when user has live location monitor renders correctly when minimized 1`] = `
<LeftPanelLiveShareWarning
  isMinimized={true}
>
  <AccessibleButton
    className="mx_LeftPanelLiveShareWarning mx_LeftPanelLiveShareWarning__minimized"
    element="div"
    onClick={[Function]}
    role="button"
    tabIndex={0}
    title="You are sharing your live location"
  >
    <div
      className="mx_AccessibleButton mx_LeftPanelLiveShareWarning mx_LeftPanelLiveShareWarning__minimized"
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      role="button"
      tabIndex={0}
      title="You are sharing your live location"
    >
      <div
        height={10}
      />
    </div>
  </AccessibleButton>
</LeftPanelLiveShareWarning>
`;

exports[`<LeftPanelLiveShareWarning /> when user has live location monitor renders correctly when not minimized 1`] = `
<LeftPanelLiveShareWarning>
  <AccessibleButton
    className="mx_LeftPanelLiveShareWarning"
    element="div"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="mx_AccessibleButton mx_LeftPanelLiveShareWarning"
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      role="button"
      tabIndex={0}
    >
      You are sharing your live location
    </div>
  </AccessibleButton>
</LeftPanelLiveShareWarning>
`;

exports[`<LeftPanelLiveShareWarning /> when user has live location monitor renders location publish error 1`] = `
<LeftPanelLiveShareWarning>
  <AccessibleButton
    className="mx_LeftPanelLiveShareWarning mx_LeftPanelLiveShareWarning__error"
    element="div"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="mx_AccessibleButton mx_LeftPanelLiveShareWarning mx_LeftPanelLiveShareWarning__error"
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      role="button"
      tabIndex={0}
    >
      An error occurred whilst sharing your live location
    </div>
  </AccessibleButton>
</LeftPanelLiveShareWarning>
`;
